clash-cores/src/Clash/Cores/Xilinx/DcFifo/Explicit.hs

```diff
@@ -0,0 +1,291 @@
+import           Clash.Explicit.Prelude
+import           Clash.Signal.Internal       (Signal (..))
+
+import           Data.Bifunctor              (second)
+import           Data.String.Interpolate     (i)
+
+-- We want constructor names to correspond 1:1 to Xilinx options it ease
+-- blackbox implementation.
+{-# HLINT ignore "Use camelCase" #-}
+
+type ResetBusy = Bit
+type Full = Bool
+type Empty = Bool
+type DataCount n = Unsigned n
+
+data ReadMode
+  = Standard_FIFO
+  | First_Word_Fall_Through
+  deriving (Show, Generic)
+
+data DcImplementation
+  = Independent_Clocks_Block_RAM
+  | Independent_Clocks_Distributed_RAM
+  | Independent_Clocks_Builtin_FIFO
+  deriving (Show, Generic)
+
+data DcConfig depth = DcConfig
+  { dcDepth          :: !depth
+  , dcImplementation :: !DcImplementation
+  , dcReadMode       :: !ReadMode
+  , dcReadDataCount  :: !Bool
+  , dcWriteDataCount :: !Bool
+  }
+  deriving (Show, Generic)
+
+defConfig :: KnownNat depth => DcConfig (SNat depth)
+defConfig = DcConfig
+  { dcDepth = SNat
+  , dcImplementation = Independent_Clocks_Block_RAM
+  , dcReadMode = Standard_FIFO
+  , dcReadDataCount = True
+  , dcWriteDataCount = True
+  }
+
+data FifoState depth n = FifoState
+  { hsQueue      :: Vec (2^depth) (Maybe (BitVector n))
+  , relativeTime :: Int
+  } deriving Show
+
+countJust :: Vec n (Maybe a) -> Int
+countJust Nil                = 0
+countJust (Cons Nothing _)   = 0
+countJust (Cons (Just _) xs) = 1 + countJust xs
+
+-- | Pop the rightmost 'Just'
+popRight :: Vec n (Maybe a) -> Maybe (a, Vec n (Maybe a))
+popRight Nil                                 = Nothing
+popRight (Cons Nothing _)                    = Nothing
+popRight (Cons (Just x) Nil)                 = Just (x, singleton Nothing)
+popRight (Cons (Just x) xs@(Cons Nothing _)) = Just (x, xs ++ singleton Nothing)
+popRight (Cons x@Just{} xs)                  = second (Cons x) <$> popRight xs
+
+data XilinxFifo read write depth n =
+  XilinxFifo
+    { writeReset :: Signal write ResetBusy
+    , isFull :: Signal write Full
+    , writeCount :: Signal write (DataCount depth)
+    , readReset :: Signal read ResetBusy
+    , isEmpty :: Signal read Empty
+    , readCount :: Signal read (DataCount depth)
+    , fifoData :: Signal read (BitVector n)
+    }
+
+-- | Mock Xilinx FIFO, see [documentation](https://docs.xilinx.com/v/u/en-US/pg057-fifo-generator)
+-- for expected behavior.
+--
+-- If 'dcReadDataCount' or 'dcWriteDataCount' are disabled, the relevant signals
+-- will be 'deepErrorX'
+dcFifo ::
+  forall depth n write read .
+  ( KnownNat n
+  , KnownDomain write
+  , KnownDomain read
+
+  , KnownNat depth
+  -- Number of elements should be between [2**4, 2**17] ~ [16, 131072].
+  , 4 <= depth
+  , depth <= 17
+  ) =>
+  DcConfig (SNat depth) ->
+
+  Clock write -> Clock read ->
+  -- | Asynchronous reset
+  Reset read ->
+
+  -- | Write data
+  Signal write (BitVector n) ->
+  -- | Write enable
+  Signal write Bool ->
```

Comment:
`Signal write (Maybe (BitVector n))`?

Reply:
With https://github.com/clash-lang/clash-compiler/pull/2202 we can do this AND get rid of the wrapper!

Reply:
Oh wow that is very nice.